In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("NbaInfo.csv")
df.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [4]:
df.columns

Index(['Name', 'Team', 'Number', 'Position', 'Age', 'Height', 'Weight',
       'College', 'Salary'],
      dtype='object')

In [5]:
df.axes

[RangeIndex(start=0, stop=458, step=1),
 Index(['Name', 'Team', 'Number', 'Position', 'Age', 'Height', 'Weight',
        'College', 'Salary'],
       dtype='object')]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


In [7]:
df.describe()

Number         Age      Weight        Salary
count  457.000000  457.000000  457.000000  4.460000e+02
mean    17.678337   26.938731  221.522976  4.842684e+06
std     15.966090    4.404016   26.368343  5.229238e+06
min      0.000000   19.000000  161.000000  3.088800e+04
25%      5.000000   24.000000  200.000000  1.044792e+06
50%     13.000000   26.000000  220.000000  2.839073e+06
75%     25.000000   30.000000  240.000000  6.500000e+06
max     99.000000   40.000000  307.000000  2.500000e+07

Inserting the data
---------------------

In [8]:
df2 = df[:]
df2["Email"] = df2["Name"] + "@gmail.com"

C:\Users\miles\AppData\Local\Temp\ipykernel_29540\1804708229.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2["Email"] = df2["Name"] + "@gmail.com"


In [9]:
df2.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary                    Email  
0              Texas  7730337.0  Avery Bradley@gmail.com  
1          Marquette  6796117.0    Jae Crowder@gmail.com  
2  Boston University        NaN   John Holland@gmail.com  
3      Georgia State  1148640.0    R.J. Hunter@gmail.com  
4                NaN  5000000.0  Jonas Jerebko@gmail.com

Here we insert the values in respective positions with loc argument
------------------ 

In [10]:
df2.insert(loc = 1, column = "WeightUser", value = 80)

In [11]:
df2.head()

Name  WeightUser            Team  Number Position   Age Height  \
0  Avery Bradley          80  Boston Celtics     0.0       PG  25.0    6-2   
1    Jae Crowder          80  Boston Celtics    99.0       SF  25.0    6-6   
2   John Holland          80  Boston Celtics    30.0       SG  27.0    6-5   
3    R.J. Hunter          80  Boston Celtics    28.0       SG  22.0    6-5   
4  Jonas Jerebko          80  Boston Celtics     8.0       PF  29.0   6-10   

   Weight            College     Salary                    Email  
0   180.0              Texas  7730337.0  Avery Bradley@gmail.com  
1   235.0          Marquette  6796117.0    Jae Crowder@gmail.com  
2   205.0  Boston University        NaN   John Holland@gmail.com  
3   185.0      Georgia State  1148640.0    R.J. Hunter@gmail.com  
4   231.0                NaN  5000000.0  Jonas Jerebko@gmail.com

In [12]:
df2["Age"] = df2["Age"] + 4
df2.head()

C:\Users\miles\AppData\Local\Temp\ipykernel_29540\1911507901.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2["Age"] = df2["Age"] + 4


Name  WeightUser            Team  Number Position   Age Height  \
0  Avery Bradley          80  Boston Celtics     0.0       PG  29.0    6-2   
1    Jae Crowder          80  Boston Celtics    99.0       SF  29.0    6-6   
2   John Holland          80  Boston Celtics    30.0       SG  31.0    6-5   
3    R.J. Hunter          80  Boston Celtics    28.0       SG  26.0    6-5   
4  Jonas Jerebko          80  Boston Celtics     8.0       PF  33.0   6-10   

   Weight            College     Salary                    Email  
0   180.0              Texas  7730337.0  Avery Bradley@gmail.com  
1   235.0          Marquette  6796117.0    Jae Crowder@gmail.com  
2   205.0  Boston University        NaN   John Holland@gmail.com  
3   185.0      Georgia State  1148640.0    R.J. Hunter@gmail.com  
4   231.0                NaN  5000000.0  Jonas Jerebko@gmail.com

while updating a column in dataframe using list comprehension
-------------------------------------------------------------

You can update a column in a pandas dataframe using list comprehension as follows:

In [13]:
df['column_name'] = [expression for item in df['column_name']]


KeyError: 'column_name'

Here, expression is the new value you want to assign to each item in the specified column. The list comprehension will iterate over each item in the specified column, apply the expression, and create a new list that replaces the original values in the column.

In [ ]:
import pandas as pd

# Create a sample dataframe
data = {'A': [1, 2, 3, 4, 5],
        'B': [10, 20, 30, 40, 50]}
df = pd.DataFrame(data)

# Update column 'B' using list comprehension
df['B'] = [val*2 for val in df['B']]

# Verify the result
print(df)

   A    B
0  1   20
1  2   40
2  3   60
3  4   80
4  5  100


In this example, the values in column 'B' are multiplied by 2 using the list comprehension, and the result is assigned back to the same column.

In [16]:
df2["Team"].value_counts()
#or
df2.value_counts(subset = "Team", normalize= True)*100

Team
New Orleans Pelicans      4.157549
Memphis Grizzlies         3.938731
New York Knicks           3.501094
Milwaukee Bucks           3.501094
Atlanta Hawks             3.282276
Utah Jazz                 3.282276
Toronto Raptors           3.282276
San Antonio Spurs         3.282276
Sacramento Kings          3.282276
Portland Trail Blazers    3.282276
Phoenix Suns              3.282276
Philadelphia 76ers        3.282276
Oklahoma City Thunder     3.282276
Boston Celtics            3.282276
Miami Heat                3.282276
Los Angeles Lakers        3.282276
Los Angeles Clippers      3.282276
Indiana Pacers            3.282276
Houston Rockets           3.282276
Golden State Warriors     3.282276
Detroit Pistons           3.282276
Denver Nuggets            3.282276
Dallas Mavericks          3.282276
Cleveland Cavaliers       3.282276
Chicago Bulls             3.282276
Charlotte Hornets         3.282276
Brooklyn Nets             3.282276
Washington Wizards        3.282276
Minnesota Timbe

In [ ]:
df2["Team"].count()

457

In [ ]:
df2["Team"].unique()

array(['Boston Celtics', 'Brooklyn Nets', 'New York Knicks',
       'Philadelphia 76ers', 'Toronto Raptors', 'Golden State Warriors',
       'Los Angeles Clippers', 'Los Angeles Lakers', 'Phoenix Suns',
       'Sacramento Kings', 'Chicago Bulls', 'Cleveland Cavaliers',
       'Detroit Pistons', 'Indiana Pacers', 'Milwaukee Bucks',
       'Dallas Mavericks', 'Houston Rockets', 'Memphis Grizzlies',
       'New Orleans Pelicans', 'San Antonio Spurs', 'Atlanta Hawks',
       'Charlotte Hornets', 'Miami Heat', 'Orlando Magic',
       'Washington Wizards', 'Denver Nuggets', 'Minnesota Timberwolves',
       'Oklahoma City Thunder', 'Portland Trail Blazers', 'Utah Jazz',
       nan], dtype=object)

In [ ]:
df2["Team"].nunique()

30

To identify null values from the dataset:
----------------------------------------

In [ ]:
df["Name"].hasnans

True

To remove the record which consist atleast 1 null value
------------------------

In [ ]:
df.dropna(how = "any") # Even if any 1 column contains null value, the record is deleted 

Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
6    Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7     Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
..             ...             ...     ...      ...   ...    ...     ...   
449    Rodney Hood       Utah Jazz     5.0       SG  23.0    6-8   206.0   
451  Chris Johnson       Utah Jazz    23.0       SF  26.0    6-6   206.0   
452     Trey Lyles       Utah Jazz    41.0       PF  20.0   6-10   234.0   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   

           College     Salary  
0            Texas  7730337.0  
1        Marquette  6796117.0  
3    Georgia State  1148640.0  
6              LSU  1170960.0  
7          Gonzaga  2165160.0  
..             ...        ...  
449           Duke  1348440.0  
451         Dayton   981348.0  
452       Kentucky  2239800.0  
453         Butler  2433333.0  
456         Kansas   947276.0  

[364 rows x 9 columns]

In [ ]:
df.dropna(how = "all")  #if all column contains null values, then the record is deleted.

Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2     John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4    Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
..             ...             ...     ...      ...   ...    ...     ...   
452     Trey Lyles       Utah Jazz    41.0       PF  20.0   6-10   234.0   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
454      Raul Neto       Utah Jazz    25.0       PG  24.0    6-1   179.0   
455   Tibor Pleiss       Utah Jazz    21.0        C  26.0    7-3   256.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   

               College     Salary  
0                Texas  7730337.0  
1            Marquette  6796117.0  
2    Boston University        NaN  
3        Georgia State  1148640.0  
4                  NaN  5000000.0  
..                 ...        ...  
452           Kentucky  2239800.0  
453             Butler  2433333.0  
454                NaN   900000.0  
455                NaN  2900000.0  
456             Kansas   947276.0  

[457 rows x 9 columns]

In [ ]:
df.dropna(subset = "College") # In this case, the specific column is mentioned, if the column College contains null, then the whole record is deleted.

Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2     John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
6    Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
..             ...             ...     ...      ...   ...    ...     ...   
449    Rodney Hood       Utah Jazz     5.0       SG  23.0    6-8   206.0   
451  Chris Johnson       Utah Jazz    23.0       SF  26.0    6-6   206.0   
452     Trey Lyles       Utah Jazz    41.0       PF  20.0   6-10   234.0   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   

               College     Salary  
0                Texas  7730337.0  
1            Marquette  6796117.0  
2    Boston University        NaN  
3        Georgia State  1148640.0  
6                  LSU  1170960.0  
..                 ...        ...  
449               Duke  1348440.0  
451             Dayton   981348.0  
452           Kentucky  2239800.0  
453             Butler  2433333.0  
456             Kansas   947276.0  

[373 rows x 9 columns]

In [ ]:
df.dropna(subset = ["College","Salary"]) 

Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
6    Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7     Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
..             ...             ...     ...      ...   ...    ...     ...   
449    Rodney Hood       Utah Jazz     5.0       SG  23.0    6-8   206.0   
451  Chris Johnson       Utah Jazz    23.0       SF  26.0    6-6   206.0   
452     Trey Lyles       Utah Jazz    41.0       PF  20.0   6-10   234.0   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   

           College     Salary  
0            Texas  7730337.0  
1        Marquette  6796117.0  
3    Georgia State  1148640.0  
6              LSU  1170960.0  
7          Gonzaga  2165160.0  
..             ...        ...  
449           Duke  1348440.0  
451         Dayton   981348.0  
452       Kentucky  2239800.0  
453         Butler  2433333.0  
456         Kansas   947276.0  

[364 rows x 9 columns]

To Fill the Nan Values in the dataset:
---------------------------------------

In [ ]:
df.fillna(0)

Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2     John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4    Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
..             ...             ...     ...      ...   ...    ...     ...   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
454      Raul Neto       Utah Jazz    25.0       PG  24.0    6-1   179.0   
455   Tibor Pleiss       Utah Jazz    21.0        C  26.0    7-3   256.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   
457              0               0     0.0        0   0.0      0     0.0   

               College     Salary  
0                Texas  7730337.0  
1            Marquette  6796117.0  
2    Boston University        0.0  
3        Georgia State  1148640.0  
4                    0  5000000.0  
..                 ...        ...  
453             Butler  2433333.0  
454                  0   900000.0  
455                  0  2900000.0  
456             Kansas   947276.0  
457                  0        0.0  

[458 rows x 9 columns]

In [ ]:
df.fillna("ABC").info() #whenever we fill the null values with any special characters, then dtype will be changed as object.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Name      458 non-null    object
 1   Team      458 non-null    object
 2   Number    458 non-null    object
 3   Position  458 non-null    object
 4   Age       458 non-null    object
 5   Height    458 non-null    object
 6   Weight    458 non-null    object
 7   College   458 non-null    object
 8   Salary    458 non-null    object
dtypes: object(9)
memory usage: 32.3+ KB


In [ ]:
df.fillna(method = 'ffill') # filled with previous values
df.fillna(method = 'bfill') # filled with next values

Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2     John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4    Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
..             ...             ...     ...      ...   ...    ...     ...   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
454      Raul Neto       Utah Jazz    25.0       PG  24.0    6-1   179.0   
455   Tibor Pleiss       Utah Jazz    21.0        C  26.0    7-3   256.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   
457    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   

               College     Salary  
0                Texas  7730337.0  
1            Marquette  6796117.0  
2    Boston University  6796117.0  
3        Georgia State  1148640.0  
4        Georgia State  5000000.0  
..                 ...        ...  
453             Butler  2433333.0  
454             Butler   900000.0  
455             Butler  2900000.0  
456             Kansas   947276.0  
457             Kansas   947276.0  

[458 rows x 9 columns]

In [ ]:
df['Salary'] = df['Salary'].fillna(0) #always fill the values with compatible datatype.

In [ ]:

df['College'] = df['College'].fillna("Texas")

In [ ]:
df["College"].hasnans

False

To Sort the dataframe based on specific column:
----------------------------------------------

In [ ]:
df.sort_values(by = "Salary", ascending = False)

Name                    Team  Number Position   Age Height  \
109      Kobe Bryant      Los Angeles Lakers    24.0       SF  37.0    6-6   
169     LeBron James     Cleveland Cavaliers    23.0       SF  31.0    6-8   
33   Carmelo Anthony         New York Knicks     7.0       SF  32.0    6-8   
251    Dwight Howard         Houston Rockets    12.0        C  30.0   6-11   
339       Chris Bosh              Miami Heat     1.0       PF  32.0   6-11   
..               ...                     ...     ...      ...   ...    ...   
269     Ray McCallum       Memphis Grizzlies     5.0       PG  24.0    6-3   
409       Greg Smith  Minnesota Timberwolves     4.0       PF  25.0   6-10   
2       John Holland          Boston Celtics    30.0       SG  27.0    6-5   
264    Jordan Farmar       Memphis Grizzlies     4.0       PG  29.0    6-2   
457              NaN                     NaN     NaN      NaN   NaN    NaN   

     Weight            College      Salary  
109   212.0              Texas  25000000.0  
169   250.0              Texas  22970500.0  
33    240.0           Syracuse  22875000.0  
251   265.0              Texas  22359364.0  
339   235.0       Georgia Tech  22192730.0  
..      ...                ...         ...  
269   190.0            Detroit         0.0  
409   250.0       Fresno State         0.0  
2     205.0  Boston University         0.0  
264   180.0               UCLA         0.0  
457     NaN              Texas         0.0  

[458 rows x 9 columns]

In [ ]:
df.sort_values(by = "Age", ascending = False) #Nan values will be at last in both asc and desc order.

Name                    Team  Number Position   Age Height  \
304    Andre Miller       San Antonio Spurs    24.0       PG  40.0    6-3   
400   Kevin Garnett  Minnesota Timberwolves    21.0       PF  40.0   6-11   
298      Tim Duncan       San Antonio Spurs    21.0        C  40.0   6-11   
261    Vince Carter       Memphis Grizzlies    15.0       SG  39.0    6-6   
102  Pablo Prigioni    Los Angeles Clippers     9.0       PG  39.0    6-3   
..              ...                     ...     ...      ...   ...    ...   
401      Tyus Jones  Minnesota Timberwolves     1.0       PG  20.0    6-2   
60   Christian Wood      Philadelphia 76ers    35.0       PF  20.0   6-11   
226   Rashad Vaughn         Milwaukee Bucks    20.0       SG  19.0    6-6   
122    Devin Booker            Phoenix Suns     1.0       SG  19.0    6-6   
457             NaN                     NaN     NaN      NaN   NaN    NaN   

     Weight         College     Salary  
304   200.0            Utah   250750.0  
400   240.0           Texas  8500000.0  
298   250.0     Wake Forest  5250000.0  
261   220.0  North Carolina  4088019.0  
102   185.0           Texas   947726.0  
..      ...             ...        ...  
401   195.0            Duke  1282080.0  
60    220.0            UNLV   525093.0  
226   202.0            UNLV  1733040.0  
122   206.0        Kentucky  2127840.0  
457     NaN           Texas        0.0  

[458 rows x 9 columns]

In [ ]:
df.sort_values(by = "Age", ascending = False, na_position = "first") # to display Nan values at top we use na_position argument in sort_values() fn.

Name                    Team  Number Position   Age Height  \
457                 NaN                     NaN     NaN      NaN   NaN    NaN   
304        Andre Miller       San Antonio Spurs    24.0       PG  40.0    6-3   
400       Kevin Garnett  Minnesota Timberwolves    21.0       PF  40.0   6-11   
298          Tim Duncan       San Antonio Spurs    21.0        C  40.0   6-11   
261        Vince Carter       Memphis Grizzlies    15.0       SG  39.0    6-6   
..                  ...                     ...     ...      ...   ...    ...   
40   Kristaps Porzingis         New York Knicks     6.0       PF  20.0    7-3   
401          Tyus Jones  Minnesota Timberwolves     1.0       PG  20.0    6-2   
60       Christian Wood      Philadelphia 76ers    35.0       PF  20.0   6-11   
226       Rashad Vaughn         Milwaukee Bucks    20.0       SG  19.0    6-6   
122        Devin Booker            Phoenix Suns     1.0       SG  19.0    6-6   

     Weight         College     Salary  
457     NaN           Texas        0.0  
304   200.0            Utah   250750.0  
400   240.0           Texas  8500000.0  
298   250.0     Wake Forest  5250000.0  
261   220.0  North Carolina  4088019.0  
..      ...             ...        ...  
40    240.0           Texas  4131720.0  
401   195.0            Duke  1282080.0  
60    220.0            UNLV   525093.0  
226   202.0            UNLV  1733040.0  
122   206.0        Kentucky  2127840.0  

[458 rows x 9 columns]

Multiple column sorting:
-----------------------

In [ ]:
df.sort_values(by = ["Team","Name"], ascending = [False,True], na_position = "first")

Name                Team  Number Position   Age Height  \
457               NaN                 NaN     NaN      NaN   NaN    NaN   
368     Alan Anderson  Washington Wizards     6.0       SG  33.0    6-6   
369      Bradley Beal  Washington Wizards     3.0       SG  22.0    6-5   
372       Drew Gooden  Washington Wizards    90.0       PF  34.0   6-10   
380    Garrett Temple  Washington Wizards    17.0       SG  30.0    6-6   
..                ...                 ...     ...      ...   ...    ...   
315      Paul Millsap       Atlanta Hawks     4.0       PF  31.0    6-8   
320   Thabo Sefolosha       Atlanta Hawks    25.0       SF  32.0    6-7   
321    Tiago Splitter       Atlanta Hawks    11.0        C  31.0   6-11   
310  Tim Hardaway Jr.       Atlanta Hawks    10.0       SG  24.0    6-6   
322    Walter Tavares       Atlanta Hawks    22.0        C  24.0    7-3   

     Weight         College      Salary  
457     NaN           Texas         0.0  
368   220.0  Michigan State   4000000.0  
369   207.0         Florida   5694674.0  
372   250.0          Kansas   3300000.0  
380   195.0             LSU   1100602.0  
..      ...             ...         ...  
315   246.0  Louisiana Tech  18671659.0  
320   220.0           Texas   4000000.0  
321   245.0           Texas   9756250.0  
310   205.0        Michigan   1304520.0  
322   260.0           Texas   1000000.0  

[458 rows x 9 columns]

If Any numerical data column  contains null value, it automatically converted to float datatype.
------------

In [ ]:
df3 = pd.read_csv("NbaInfo.csv")
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


Type Casting:
-------------

In [ ]:
df['Salary'] = df['Salary'].astype(int) #float to int, type casting is possible only if new datatype is compatible with the old one.

In [ ]:
df['Age'].fillna(method="ffill",inplace=True)
df['Age'] = df['Age'].astype(int)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       458 non-null    int32  
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   458 non-null    object 
 8   Salary    458 non-null    int32  
dtypes: float64(2), int32(2), object(5)
memory usage: 28.8+ KB


Handling Duplicate Records
----------------

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       458 non-null    int32  
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   458 non-null    object 
 8   Salary    458 non-null    int32  
dtypes: float64(2), int32(2), object(5)
memory usage: 28.8+ KB


In [ ]:
df['Team'] = df['Team'].astype('category')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   Name      457 non-null    object  
 1   Team      457 non-null    category
 2   Number    457 non-null    float64 
 3   Position  457 non-null    object  
 4   Age       458 non-null    int32   
 5   Height    457 non-null    object  
 6   Weight    457 non-null    float64 
 7   College   458 non-null    object  
 8   Salary    458 non-null    int32   
dtypes: category(1), float64(2), int32(2), object(4)
memory usage: 26.9+ KB


In the above case, 28.8 kb has been reduced to 26.5 kb using category datatype.
-----------------------

In [ ]:
for i in df.columns:
    print(f'{i}, {df[i].nunique()} ')

    # we can infer position has only 5 categories, so if change the datatype to category, the memory size can be optimized.

Name, 457 
Team, 30 
Number, 53 
Position, 5 
Age, 22 
Height, 18 
Weight, 87 
College, 118 
Salary, 310 


In [ ]:
df['Position'] = df['Position'].astype('category')
df.info() #still it has been reduced to 24 KB.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   Name      457 non-null    object  
 1   Team      457 non-null    category
 2   Number    457 non-null    float64 
 3   Position  457 non-null    category
 4   Age       458 non-null    int32   
 5   Height    457 non-null    object  
 6   Weight    457 non-null    float64 
 7   College   458 non-null    object  
 8   Salary    458 non-null    int32   
dtypes: category(2), float64(2), int32(2), object(3)
memory usage: 24.0+ KB
